### ISSUE TITLE
ArrayTrailingComma: Extra coma is required in multiline array value

### ISSUE THREAD

**Issue by romani** [checkstyle/checkstyle#1509]
from MainTest.java of checkstyle source

```
        // we just reference there all violations
        final String[][] outputValues = new String[][]{
            {"ClassCouplingCheckTestInput", "InnerClass", "7:19"},
            {"BooleanExpressionComplexityCheckTestInput",
                "BooleanExpressionComplexityCheckTestInput", "3:14", },
            {"BooleanExpressionComplexityCheckTestInput", "Settings", "53:19"},
        };
```

veird part is ""3:14", }," checkstyle demand that coma.
Config is checkstyle_checks.xml 
##

**Comment by Vladlis** [comment]
As it can be seen from test, this behaviour is expected: [input](https://github.com/Vladlis/checkstyle/blob/i1509-trailing-co/src/test/resources/com/puppycrawl/tools/checkstyle/InputArrayTrailingComma.java#L34) and [test](https://github.com/Vladlis/checkstyle/blob/i1509-trailing-co/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/ArrayTrailingCommaCheckTest.java#L39).
From documentation:
> The check allows leaving out the comma at the end if both the left and right curly brackets are on the same line.

the same refers to the array values - if they are multiline, they should contain a trailing comma.

**Comment by mkordas** [comment]
For me it is still  not intuitive. Training comma brings value usually only if both braces are on different lines. 

My proposition is to change the rule to behave as follows:
The check allows leaving out the comma at the end if **either** the left **or** right curly brackets are on the same line.

**Comment by romani** [comment]
Intention of this rule is not clear for me at all, but that logic exists for many years and nobody complained about that, and it is clearly documented. That is why I decided to close that issue.

I could connect to author of this update, to get from him his motivation of doing that.

**Comment by romani** [comment]
>The check allows leaving out the comma at the end if either the left or right curly brackets are on the same line.

I do not understand meaning of that rule, please rephrase for me and lets start discussion by code examples (with violation and not)

**Comment by mkordas** [comment]
@romani - sure.

Main benefit of training comma is that when you add new entry to array, no surrounding lines are changed:
```java
{
    100000000000000000000, 
    200000000000000000000, // OK
}
```
```java
{
    100000000000000000000,
    200000000000000000000,
    300000000000000000000,  // Just this line added, no other changes
}
```
If closing brace is on the same line as training comma, this benefit is gone:
```java
{100000000000000000000, 
    200000000000000000000,} // Trailing comma not needed, line needs to be modified anyway

```
```java
{100000000000000000000, 
    200000000000000000000, // Modified line
    300000000000000000000,} // Added line
```
If opening brace is on the same line as training comma there's also (more arguable) problem:
```java
{100000000000000000000, // Line cannot be just duplicated to slightly modify entry
} 
```
```java
{100000000000000000000,
    100000000000000000001, // More work needed to duplicate
} 
```

**Comment by romani** [comment]
@mkordas 

> If opening brace is on the same line as training comma there's also (more arguable) problem:

I do not see changes to first line, I do not understand how "{" location affect the rule for "," presence. Please extend examples for this case.

**Comment by romani** [comment]
@os97673 , looks like you were an original author or this rule https://github.com/checkstyle/checkstyle/blame/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ArrayTrailingCommaCheck.java#L25 , it would be so kind if you can bring from a history, of 2003 year :) , why "," is required (please see issue description).

**Comment by os97673** [comment]
Well, the check is supposed to be used when you either write one-line arrays
```java
{ 1, 2, 3}
```
or place keep curly braces on separate lines and one element per line
```java
{
  1,
  2,
  3,
}
```
if you do not require this then the check is (almost) useless.
So the check should be always used in combination with another one (for array formatting).
Thus I'd keep the check as is and do suggest do not use it in some cases :)
Hope I've answered the question ;)

**Comment by mkordas** [comment]
@romani 
> I do not see changes to first line, I do not understand how "{" location affect the rule for "," presence. Please extend examples for this case.

Example I've given should be enough. If `{` is on the same line as `,` you have more work to duplicate the line. Instead of duplicating and changing just value, you also need to remove `{` from duplicated line.

**Comment by rnveach** [comment]
Fix is merged
